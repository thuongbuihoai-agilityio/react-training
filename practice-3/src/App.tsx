import Category from "./components/Categories/Categories"
import Footer from "./components/Footer/Footer"
import Header from "./components/Header/Header"
import Navigation from "./components/Navigation/Navigation"
import ViewProductItem from "./components/ViewsProducts/ViewProductItem/ViewProductsItem"
import ProductDetails from "./pages/ProductDetail/ProductDetail"
import ProductItem from "./pages/ProductItem/ProductItem"

function App() {
  return (
    <>
      <Navigation />
      <Header />
      <Category />
      <ViewProductItem />
      <ProductItem />
      <ProductDetails />
      <Footer />
    </>
  )
}

export default App
