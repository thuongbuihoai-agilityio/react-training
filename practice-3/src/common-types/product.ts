export interface Product {
  id: string;
  name: string;
  price: number;
  images: string[];
  quantity: number;
  categoryId: string;
  description: string;
  product?: {};
}

export interface ProductItemProps {
  product: Product;
  deleteProduct: (id: string) => void;
}

export interface ProductCardProps {
  product: Product;
}

export interface ProductContext {
  products: Product[];
  dispatch: Function,
  searchValue: string;
  setSearchValue: Function;
}