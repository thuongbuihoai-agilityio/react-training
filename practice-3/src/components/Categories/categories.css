.categories {
  margin-top: 50px;
  text-align: center;
}

.categories__title {
  text-transform: uppercase;
  font-weight: bold;
}

.categories__list {
  display: flex;
  justify-content: center;
  margin: 30px 0 40px 0;
  padding: 0;
}

.categories__item {
  list-style: none;
  cursor: pointer;
  margin: 0 30px;
  position: relative;
  color: #696969;
}

.categories__item:hover {
  color: #ebb576;
}

.categories__item::after {
  content: "";
  width: 5px;
  height: 5px;
  border-radius: 50%;
  background-color: #555;
  position: absolute;
  right: -30px;
  top: 45%;
  color: #696969;
}

.categories__item > a {
  color: black;
  text-decoration: none;
}