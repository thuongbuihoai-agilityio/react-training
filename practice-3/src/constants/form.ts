import { RULES } from "./rules"

const FORM_VALUES = {
  categoryId: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  name: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  price: {
    value: "",
    rules: [RULES.REQUIRED, RULES.NUMBER],
    error: "",
  },
  quantity: {
    value: "",
    rules: [RULES.REQUIRED, RULES.NUMBER],
    error: "",
  },
  description: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  images: {
    value: "",
    rules: [],
    error: "",
  }
}

export default FORM_VALUES
