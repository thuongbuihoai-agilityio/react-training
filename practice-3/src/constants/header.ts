const USERS = {
  username: "Hoai Thuong",
};

export { USERS }