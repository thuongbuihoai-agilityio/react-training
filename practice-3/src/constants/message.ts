const SUCCESS_MSG = {
  MESSAGE_ADD_PRODUCT: "Product add successful",
  MESSAGE_UPDATE_PRODUCT: "Product update successful",
  MESSAGE_DELETE_PRODUCT: "Product delete successful",
};

export { SUCCESS_MSG };
