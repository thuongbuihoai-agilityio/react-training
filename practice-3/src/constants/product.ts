import { Product } from "types/product";

const PRODUCT_CONTENT: Product = {
  id: "1",
  categoryId: "1651924356334",
  name: "Cheese cake",
  price: 20,
  quantity: 5,
  description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
  images: [
    "/src/assets/images/products/cheese-cake.jpg",
    "/src/assets/images/products/cheese-cake.jpg",
    "/src/assets/images/products/cheese-cake.jpg"
  ]
}

export { PRODUCT_CONTENT };
