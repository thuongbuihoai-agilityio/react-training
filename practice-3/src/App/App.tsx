import React from "react";
import { memo } from "react";
import { SWRConfig } from "swr";
import { Route, Routes } from "react-router-dom";
import "react-toastify/dist/ReactToastify.css";
import { ToastContainer } from "react-toastify";
import Categories from "@/components/Categories/Categories";
import Footer from "@/components/common/Footer/Footer";
import Header from "@/components/common/Header/Header";
import Navigation from "@/components/common/Navigation/Navigation";
import ProductDetails from "@/pages/ProductDetail/ProductDetail";
import ProductListView from "@/pages/ProductListView/ProductListView";
import ProductGridView from "@/pages/ProductGridView/ProductGridView";
import AppProvider from "@/context/AppContext";

const App: React.FC = memo(() => {
  return (
    <SWRConfig value={{ revalidateOnFocus: false }}>
      <ToastContainer />
      <Header />
      <AppProvider>
        <Navigation />
        <Categories />
        <Routes>
          <Route path="/" element={<ProductListView />} />
          <Route path="/products" element={<ProductGridView />} />
          <Route path={`/product/:id`} element={<ProductDetails />} />
        </Routes>
      </AppProvider>
      <Footer />
    </SWRConfig>
  );
});

export default App;
