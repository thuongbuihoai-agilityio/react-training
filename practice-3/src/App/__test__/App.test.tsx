import { useState } from "react";
import {render, screen} from "@testing-library/react";
import {createMemoryHistory} from "history";
import { Router } from "react-router-dom";
import "@testing-library/jest-dom";
import { PRODUCT_MOCKING_LIST } from "@/constants/product";
import { CATEGORY_MOCKING_LIST } from "@/constants/categories";
import App from "../App";

const useProductsMock = { products: PRODUCT_MOCKING_LIST };
  jest.mock("../../hooks/useProducts.ts", () => ({
    default: jest.fn(() => useProductsMock),
}));

const useCategoriesMock = { categories: CATEGORY_MOCKING_LIST };
  jest.mock("../../hooks/useCategories.ts", () => ({
    default: jest.fn(() => useCategoriesMock),
}));

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

describe("App component", () => {
  beforeEach(()=>{
    (useState as jest.Mock).mockImplementation(jest.requireActual("react").useState);
  })

  test("rendering a view product list component that uses useLocation", () => {
    const history = createMemoryHistory();
    const route = "/";
    history.push(route);
    render(
      <Router location={history.location} navigator={history}>
        <App />
      </Router>
    )
    expect(screen.getByTestId("view-product-list")).toBeInTheDocument();
  })

  test("rendering a product list component that uses useLocation", () => {
    const history = createMemoryHistory();
    const route = "/products";
    history.push(route);
    render(
      <Router location={history.location} navigator={history}>
        <App />
      </Router>
    )
    expect(screen.getByTestId("product-list-page")).toBeInTheDocument();
  });

  test("rendering a product detail component that uses useLocation", () => {
    const history = createMemoryHistory();
    const route = "/product/:id";
    history.push(route);
    render(
      <Router location={history.location} navigator={history}>
        <App />
      </Router>
    )
    expect(screen.getByTestId("product-detail-page")).toBeInTheDocument();
  })

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Router location={history.location} navigator={history}>
        <App />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
})