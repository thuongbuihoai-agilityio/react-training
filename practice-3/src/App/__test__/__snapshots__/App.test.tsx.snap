// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="header"
    data-testid="header"
  >
    <div
      class="header__logo"
    >
      <a
        data-testid="logo"
        href="#Home"
      >
        <img
          class="logo"
        />
      </a>
    </div>
    <div
      class="header__about"
    >
      <div
        class="menu"
        data-testid="menu"
      >
        <ul
          class="menu__list"
        >
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              Home
            </a>
          </li>
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              About
            </a>
          </li>
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              Shop
            </a>
          </li>
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              Blog
            </a>
          </li>
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              FAQ
            </a>
          </li>
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              Contact
            </a>
          </li>
        </ul>
      </div>
      <button
        class="btn btn__btn btn__primary"
      >
        ORDER ONLINE
      </button>
    </div>
  </header>
  <figure>
    <img
      alt=""
      class="header__img"
    />
  </figure>
  <nav
    class="nav"
    data-testid="navigation"
  >
    <ul
      class="nav__list"
    >
      <li
        class="nav__item"
      >
        <a
          class="nav__link"
          href=""
        >
           
          <i
            class="fa fa-clock"
          />
           Mon - Sat: 8AM - 5PM
        </a>
      </li>
      <li
        class="nav__item"
      >
        <a
          class="nav__link"
          href=""
        >
           
          <i
            class="fa fa-phone"
          />
           123 456 7890
        </a>
      </li>
    </ul>
    <div
      class="search"
      data-testid="input-search"
    >
      <input
        class="search__text"
        placeholder="Search item"
        type="text"
      />
    </div>
  </nav>
  <div
    class="categories"
    data-testid="categories"
  >
    <p
      class="categories__title"
    >
      What are you looking for here?
    </p>
    <ul
      class="categories__list"
    >
      <li
        class="categories__item active"
        data-testid="category-item"
      >
        All
      </li>
      <li
        class="categories__item inactive"
        data-index="1"
        data-testid="category-item"
      >
        Cake
      </li>
      <li
        class="categories__item inactive"
        data-index="2"
        data-testid="category-item"
      >
        Breads
      </li>
    </ul>
  </div>
  <div
    class="viewProduct__list"
    data-testid="view-product-list"
  >
    <div
      class="viewProduct__item"
    >
      <div
        class="productViewPage"
        data-testid="productViewPage"
      >
        <img
          class="productViewPage__image"
          src="/src/assets/images/products/cheese-cake.jpg"
        />
        <div
          class="productViewPage__content"
        >
          <a
            class="productViewPage__link"
            href="/product/1"
          >
            <h2
              class="productViewPage__title"
              data-testid="title-test"
            >
              Cheese cake
            </h2>
          </a>
          <p
            class="productViewPage__description"
            data-testid="text"
          >
            Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.
          </p>
          <p
            class="productViewPage__price"
            data-testid="price"
          >
            $20
          </p>
        </div>
      </div>
    </div>
    <div
      class="viewProduct__item"
    >
      <div
        class="productViewPage"
        data-testid="productViewPage"
      >
        <img
          class="productViewPage__image"
          src="/src/assets/images/products/cheese-cake.jpg"
        />
        <div
          class="productViewPage__content"
        >
          <a
            class="productViewPage__link"
            href="/product/2"
          >
            <h2
              class="productViewPage__title"
              data-testid="title-test"
            >
              Amaretti Cookie
            </h2>
          </a>
          <p
            class="productViewPage__description"
            data-testid="text"
          >
            Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.
          </p>
          <p
            class="productViewPage__price"
            data-testid="price"
          >
            $10
          </p>
        </div>
      </div>
    </div>
  </div>
  <a
    class="viewProduct__link"
    href="/products"
  >
    <button
      class="btn btn__btn btn__secondary"
    >
      VIEW ALL PRODUCTS
    </button>
  </a>
  <footer
    class="footer"
  >
    <p
      class="footer__contact"
      data-testid="footer"
    >
      Copyright © 2022 —  Bakery WordPress Theme by Milingona.
    </p>
  </footer>
</DocumentFragment>
`;
